lib/Jsrt/JsrtDiag.cpp

```diff
@@ -0,0 +1,688 @@
+        JsrtDebugManager* jsrtDebugManager = currentContext->GetRuntime()->GetJsrtDebugManager();
+
+        VALIDATE_IS_DEBUGGING(jsrtDebugManager);
+
+        Js::JavascriptArray* scripts = jsrtDebugManager->GetScripts(scriptContext);
+
+        if (scripts != nullptr)
+        {
+            *scriptsArray = scripts;
+            return JsNoError;
+        }
+
+        return JsErrorDiagUnableToPerformAction;
+    });
+}
+
+CHAKRA_API JsDiagGetSource(
+    _In_ unsigned int scriptId,
+    _Out_ JsValueRef *source)
+{
+    return ContextAPIWrapper<false>([&](Js::ScriptContext *scriptContext) -> JsErrorCode {
+
+        PARAM_NOT_NULL(source);
+
+        *source = JS_INVALID_REFERENCE;
+
+        JsrtContext *currentContext = JsrtContext::GetCurrent();
+
+        JsrtDebugManager* jsrtDebugManager = currentContext->GetRuntime()->GetJsrtDebugManager();
+
+        VALIDATE_IS_DEBUGGING(jsrtDebugManager);
+
+        Js::DynamicObject* sourceObject = jsrtDebugManager->GetSource(scriptContext, scriptId);
+
+        if (sourceObject != nullptr)
+        {
+            *source = sourceObject;
+            return JsNoError;
+        }
+
+        return JsErrorInvalidArgument;
+    });
+}
+
+CHAKRA_API JsDiagRequestAsyncBreak(
+    _In_ JsRuntimeHandle runtimeHandle)
+{
+    return GlobalAPIWrapper([&]() -> JsErrorCode {
+
+        VALIDATE_INCOMING_RUNTIME_HANDLE(runtimeHandle);
+
+        JsrtRuntime * runtime = JsrtRuntime::FromHandle(runtimeHandle);
+
+        JsrtDebugManager* jsrtDebugManager = runtime->GetJsrtDebugManager();
+
+        VALIDATE_IS_DEBUGGING(jsrtDebugManager);
+
+        for (Js::ScriptContext *scriptContext = runtime->GetThreadContext()->GetScriptContextList();
+        scriptContext != nullptr && !scriptContext->IsClosed();
+            scriptContext = scriptContext->next)
+        {
+            jsrtDebugManager->EnableAsyncBreak(scriptContext);
+        }
+
+        return JsNoError;
+    });
+}
+
+CHAKRA_API JsDiagGetBreakpoints(
+    _Out_ JsValueRef *breakpoints)
+{
+    return GlobalAPIWrapper([&]() -> JsErrorCode {
+
+        PARAM_NOT_NULL(breakpoints);
+
+        *breakpoints = JS_INVALID_REFERENCE;
+
+        JsrtContext *currentContext = JsrtContext::GetCurrent();
+
+        Js::JavascriptArray* bpsArray = currentContext->GetScriptContext()->GetLibrary()->CreateArray();
+
+        JsrtRuntime * runtime = currentContext->GetRuntime();
+
+        ThreadContextScope scope(runtime->GetThreadContext());
+
+        if (!scope.IsValid())
+        {
+            return JsErrorWrongThread;
+        }
+
+        JsrtDebugManager* jsrtDebugManager = runtime->GetJsrtDebugManager();
+
+        VALIDATE_IS_DEBUGGING(jsrtDebugManager);
+
+        for (Js::ScriptContext *scriptContext = runtime->GetThreadContext()->GetScriptContextList();
+        scriptContext != nullptr && !scriptContext->IsClosed();
+            scriptContext = scriptContext->next)
+        {
+            jsrtDebugManager->GetBreakpoints(&bpsArray, scriptContext);
```

Comment:
oh I see

---
In reply to: [62733368](https://github.com/Microsoft/ChakraCore/pull/926#discussion_r62733368) [](ancestors = 62733368,62732695)